subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_espi.c

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2022 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/espi.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/mgmt/ec_host_cmd/ec_host_cmd.h>
+#include <zephyr/mgmt/ec_host_cmd/backend.h>
+
+LOG_MODULE_REGISTER(host_cmd_espi, CONFIG_EC_HC_LOG_LEVEL);
+
+#define RX_HEADER_SIZE (sizeof(struct ec_host_cmd_request_header))
+
+struct ec_host_cmd_espi {
+	/* eSPI device instance */
+	const struct device *espi_dev;
+	/* Context for read operation */
+	struct ec_host_cmd_rx_ctx *rx_ctx;
+	/* Transmit buffer */
+	struct ec_host_cmd_tx_buf *tx;
+	/* eSPI callback */
+	struct espi_callback espi_cb;
+};
+
+#define EC_HOST_CMD_ESPI_DEFINE(_name)                                                             \
+	static struct ec_host_cmd_espi _name##_hc_espi;                                            \
+	struct ec_host_cmd_backend _name = {                                                       \
+		.api = &ec_host_cmd_api,                                                           \
+		.ctx = (struct ec_host_cmd_espi *)&_name##_hc_espi,                                \
+	}
+
+static void espi_handler(const struct device *dev, struct espi_callback *cb,
+			 struct espi_event espi_evt)
+{
+	struct ec_host_cmd_espi *hc_espi = CONTAINER_OF(cb, struct ec_host_cmd_espi, espi_cb);
+	uint16_t event_type = (uint16_t)espi_evt.evt_details;
+	/* tx stores the shared memory buf pointer and size, so use it */
+	const struct ec_host_cmd_request_header *rx_header = hc_espi->tx->buf;
+	const size_t shared_size = hc_espi->tx->len_max;
+	const uint16_t rx_valid_data_size = rx_header->data_len + RX_HEADER_SIZE;
+
+	if (event_type != ESPI_PERIPHERAL_EC_HOST_CMD) {
+		return;
+	}
+
+	/* Only support version 3 and make sure the number of bytes to copy is not
+	 * bigger than rx buf size or the shared memory size
+	 */
+	if (rx_header->prtcl_ver != 3 ||
+	    rx_valid_data_size > CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER ||
+	    rx_valid_data_size > shared_size) {
```

Comment:
Aren't these 2 values equal?

Reply:
No, `rx_valid_data_size` is amount of bytes to copy based on a header, `CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER` is the rx buffer size where data are copied to, `shared_size` is a size of the eSPI memory, where data are copied from.